### PR TITLE
Enable inference mode for testing and predicting

### PR DESCRIPTION
**Pull request by tangbinh**
## What does this PR do?

* As proposed in #8499, we enable [inference mode](https://pytorch.org/docs/1.9.0/generated/torch.autograd.inference_mode.html?highlight=inference_mode#torch.autograd.inference_mode) for evaluation and prediction. The feature was included recently in PyTorch 1.9 and comes with computational benefits.
* We also correct a link to the previous pull request #8685 in the change log.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -1094,8 +1094,9 @@ def _run_evaluate(self) -> _EVALUATE_OUTPUT:
         # reset trainer on this loop and all child loops in case user connected a custom loop
         self._evaluation_loop.trainer = self
 
-        with self.profiler.profile(f"run_{self.state.stage}_evaluation"), torch.no_grad():
-            eval_loop_results = self._evaluation_loop.run()
+        with self.profiler.profile(f"run_{self.state.stage}_evaluation"):
+            with torch.inference_mode() if hasattr(torch, "inference_mode") else torch.no_grad():
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:1098]: - Check the PyTorch version explicitly for when inference mode was made available. That should be safer than relying on `hasattr` checks. you could split this out to a separate utility (e.g. `_is_inference_mode_available`) so it can be shared across loops
- Evaluation also includes validation, in addition to testing. Since validation runs during fitting as well, I'm not sure how sae it is to include inference mode if we're fitting. 
To be safe, you could check that the trainer is not fitting before enabling this

**Comment by ananthsub** [comment]
@tangbinh thanks for working on this! It'd be awesome to include information on performance speedups for some candidate models as well

**Comment by tangbinh** [comment]
Should we allow the users to opt out of the inference mode if they want to? I imagine some cases where the gradients are needed at test time (e.g. higher-order derivatives), and being able to turn it off by a flag might be important. I'm think of having `_TORCH_GREATER_EQUAL_1_9 and os.getenv("NO_INFERENCE_MODE", "False").lower() == "false"`.

**Comment by tangbinh** [comment]
> @tangbinh ... It'd be awesome to include information on performance speedups for some candidate models as well

Do you have suggestions for the candidate models and some datasets to test them on? I was running some simple tasks and didn't notice much of a difference.

**Comment by ananthsub** [comment]
> Do you have suggestions for the candidate models and some datasets to test them on? I was running some simple tasks and didn't notice much of a difference.

@Borda do you recommend the models used for the parity benchmark?

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:1098]: And let's move this to its own function decorated with `contextmanager` for readability.

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -1110,7 +1111,7 @@ def _run_predict(self) -> Optional[_PREDICT_OUTPUT]:
         self.reset_predict_dataloader(self.lightning_module)
         # reset trainer on this loop and all child loops in case user connected a custom loop
         self.predict_loop.trainer = self
-        with torch.no_grad():
+        with torch.inference_mode() if hasattr(torch, "inference_mode") else torch.no_grad():
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:1114]: Same here.

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -1334,3 +1335,11 @@ def _on_exception(self):
         # save a checkpoint for fault tolerant training. we don't use `log_dir` to minimize the chances of failure.
         file_path = os.path.join(self.default_root_dir, ".pl_auto_save.ckpt")
         self.save_checkpoint(file_path)
+
+    @contextmanager
+    def evaluation_context(self) -> Generator:
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:1340]: would we also be happy with making this a utility function?
And I currently see no reason to include the check of `trainer.state.fn`.

Comment (tangbinh) [reply pytorch_lightning/trainer/trainer.py:1340]: @ananthsub Do you think we should move this utility function somewhere else for reusability? In that case, should we leave the check against `fit` and `validation` loops here?

Comment (ananthsub) [reply pytorch_lightning/trainer/trainer.py:1340]: @awaelchli @carmocca if we want to enable inference mode for validation in addition to testing and prediction, this becomes a super simple utility function.

I haven't found a good example for why this wouldn't work for validation, given we have no_grad already set.

One thing we need to allow for is users disabling this mode for any reason. To that end, we should document that this mode will be turned on by default for validation/testing/prediction.

Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:1340]: From the docs:

> Enable inference mode when you are performing computations that don’t need to be recorded in the backward graph, AND you don’t plan on using the tensors created in inference mode in any computation that is to be recorded by autograd later.

This constraint is not required by `no_grad`.

I also can't come up with a good example that breaks this constraint during fitting validation, but some users might need to

**Review by carmocca — COMMENTED** [review]
For the 2 cases where `no_grad` has been replaced already, we don't need to check the `TrainerFn` because by design, they will only be called during `trainer.validate()`, `trainer.test()`, or `trainer.predict`. Where no grad operations should be enabled.

This is different to the following, which is the validation part of fitting, which is where I'm not sure if `no_grad` should be kept:

https://github.com/PyTorchLightning/pytorch-lightning/blob/e0605472306d6b95bf2616ab88f8c29f4498402e/pytorch_lightning/loops/epoch/training_epoch_loop.py#L247

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/trainer.py:1340]:
```suggestion
    def _evaluation_context(self) -> Generator:
```

prefer keeping this private in the meantime.

For now I think this is the safest route. If we also want to enable this for validation, we easily can. Are there docs we should update in case users see any side effects with inference mode and need this disabled?

Comment (tangbinh) [reply pytorch_lightning/trainer/trainer.py:1340]: @carmocca @ananthsub What do you think about this? Is there a better way to let users override the inference mode?

```python
@contextmanager
def _evaluation_context(self) -> Generator:
    with torch.inference_mode() if _TORCH_GREATER_EQUAL_1_9 and os.getenv(
        "NO_INFERENCE_MODE", "False"
    ).lower() == "false" and self.state.fn in (
        TrainerFn.TESTING,
        TrainerFn.PREDICTING,
    ) else torch.no_grad():
        yield
```

Comment (ananthsub) [reply pytorch_lightning/trainer/trainer.py:1340]: @tangbinh I don't think we need a flag. we don't offer a similar one for no_grad. users are free to specify
```
with torch.inference_mode(mode=False):
    ....
```
inside of their code that needs to disable this

**Comment by tangbinh** [comment]
@ananthsub  @tchaton @awaelchli @carmocca  Please see [this notebook](https://colab.research.google.com/drive/1GysINpTTt9g55r3XG4z3vdW458sopO7W?usp=sharing) for some performance tests. It looks like for some common models and datasets, enabling inference mode doesn't make much of a difference in terms of inference time. The observation is consistent on both GPU and CPU. That said, I think it doesn't hurt if we decide to move forward with `inference mode`.

**Review by ananthsub — APPROVED** [review]
thanks for working on this @tangbinh !

File: pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -1396,3 +1397,11 @@ def _on_exception(self):
         # save a checkpoint for fault tolerant training. we don't use `log_dir` to minimize the chances of failure.
         file_path = os.path.join(self.default_root_dir, ".pl_auto_save.ckpt")
         self.save_checkpoint(file_path)
+
+    @contextmanager
+    def _evaluation_context(self) -> Generator:
+        with torch.inference_mode() if _TORCH_GREATER_EQUAL_1_9 and self.state.fn in (
+            TrainerFn.TESTING,
+            TrainerFn.PREDICTING,
+        ) else torch.no_grad():
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:1406]:
```suggestion
        with torch.inference_mode() if _TORCH_GREATER_EQUAL_1_9 else torch.no_grad():
```
I don't see a reason to limit this to test and predict only. 
Should we move this context manager out of the trainer? It does not depend on the state, only the pytorch version.

Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:1406]: > I don't see a reason to limit this to test and predict only.

+1, see my previous comment: https://github.com/PyTorchLightning/pytorch-lightning/pull/8813#pullrequestreview-730040781

Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:1406]: +1 for your previous comment, I did not see it

**Comment by tchaton** [comment]
Dear @tangbinh @ananthsub,

Could we get the benchmark for some traditional models to validate this doesn't bring a regression for Lightning performances + udpated docs.

Best,
T.C

File: pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -1396,3 +1397,11 @@ def _on_exception(self):
         # save a checkpoint for fault tolerant training. we don't use `log_dir` to minimize the chances of failure.
         file_path = os.path.join(self.default_root_dir, ".pl_auto_save.ckpt")
         self.save_checkpoint(file_path)
+
+    @contextmanager
+    def _evaluation_context(self) -> Generator:
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:1402]: Should this be owned by the loop instead ?

File: pytorch_lightning/trainer/trainer.py (hunk 4 above)
Comment (tangbinh) [reply pytorch_lightning/trainer/trainer.py:1406]: > Could we get the benchmark for some traditional models to validate this doesn't bring a regression for Lightning performances + updated docs.

Would you mind proposing some traditional models that you have in mind? I've done [some benchmark tests](https://colab.research.google.com/drive/1GysINpTTt9g55r3XG4z3vdW458sopO7W?usp=sharing), although they don't follow previous examples as I couldn't find any good one.

**Comment by tangbinh** [comment]
> Could we get the benchmark for some traditional models to validate this doesn't bring a regression for Lightning performances + updated docs.

Would you mind proposing some traditional models that you have in mind? I've done [some benchmark tests](https://colab.research.google.com/drive/1GysINpTTt9g55r3XG4z3vdW458sopO7W?usp=sharing), although they don't follow previous examples as I couldn't find any good one.

File: pytorch_lightning/trainer/trainer.py (hunk 4 above)
Comment (tangbinh) [reply pytorch_lightning/trainer/trainer.py:1406]: @carmocca  @awaelchli Okay, we can remove the `self.state.fn` check. @ananthsub and I were just a bit more cautious about turning on the inference mode during validation.

File: pytorch_lightning/trainer/trainer.py (hunk 5 above)
Comment (tangbinh) [reply pytorch_lightning/trainer/trainer.py:1402]: @tchaton Are you suggesting that we should move this context to `evaluation_epoch_loop.py`? I'm just not sure where's best to put it.

File: pytorch_lightning/trainer/trainer.py (hunk 4 above)
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:1406]: @tangbinh keep in mind the `trainer.fit` validation and the `trainer.validate` validation used different code paths.

In this PR you are just applyint the nee context manager to the latter

File: pytorch_lightning/trainer/trainer.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:1402]: Yes, I think this could be in the scope of the loops more than the trainer.

File: pytorch_lightning/trainer/trainer.py (hunk 4 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/trainer.py:1406]:
```suggestion
        with torch.inference_mode() if _TORCH_GREATER_EQUAL_1_9 else torch.no_grad():
```

**Review by carmocca — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 5 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/trainer.py:1402]: @tchaton @awaelchli - does this mean the FitLoop needs to override `run` ? if so, that introduces a risk that the fit loop override doesn't stay in sync with the base Loop `run` definition.

asking since anomaly detection is taking a very similar route as what's provided here: https://github.com/PyTorchLightning/pytorch-lightning/pull/9175
